### PR TITLE
draft: Introduce protobuf schemas for seqvar query output

### PR DESCRIPTION
**Pull request by jakibaki**
Still very WIP, just opening to track progress ^^

Will solve #375

### CODE REVIEW

File: protos/varfish/v1/query_seqvars.proto [hunk 1]
```diff
@@ -0,0 +1,247 @@
+syntax = "proto3";
+
+package varfish.v1.seqvars;
+
+// Suggestion from emily that might be worth discussing: a few members where I think the enabled bool can be implied by the presence/nonpresence of an optional submessage  making for cleaner code & less branches
+
+
+
+enum Consequence {
```
Comment (holtgrewe) [inline protos/varfish/v1/query_seqvars.proto:9]:
```
class SeqvarsVariantConsequenceChoice(str, Enum):
    """The variant consequence."""

    # high impact

    # skipped chromosome_number_variation
    # skipped exon_loss_variant
    #: Frameshift variant
    FRAMESHIFT_VARIANT = "frameshift_variant"
    #: Rare amino acid variant
    RARE_AMINO_ACID_VARIANT = "rare_amino_acid_variant"
    #: Splice acceptor variant
    SPLICE_ACCEPTOR_VARIANT = "splice_acceptor_variant"
    #: Splice donor variant
    SPLICE_DONOR_VARIANT = "splice_donor_variant"
    #: Start lost
    START_LOST = "start_lost"
    #: Stop gained
    STOP_GAINED = "stop_gained"
    #: Stop lost
    STOP_LOST = "stop_lost"
    # skipped transcript_ablation

    # moderate impact

    # 3' UTR truncation
    THREE_PRIME_UTR_TRUNCATION = "3_prime_UTR_truncation"
    # 5' UTR truncation
    FIVE_PRIME_UTR_TRUNCATION = "5_prime_UTR_truncation"
    #: Conservative inframe deletion
    CONSERVATIVE_INFRAME_DELETION = "conservative_inframe_deletion"
    #: Conservative inframe insertion
    CONSERVATIVE_INFRAME_INSERTION = "conservative_inframe_insertion"
    #: Disruptive inframe deletion
    DISRUPTIVE_INFRAME_DELETION = "disruptive_inframe_deletion"
    #: Disruptive inframe insertion
    DISRUPTIVE_INFRAME_INSERTION = "disruptive_inframe_insertion"
    #: Missense variant
    MISSENSE_VARIANT = "missense_variant"
    #: skipped regulatory_region_ablation
    #: Splice region variant
    SPLICE_REGION_VARIANT = "splice_region_variant"
    # skipped TFBS_ablation

    # low impact

    # skipped 5_prime_UTR_premature_start_codon_gain_variant
    #: Initiator codon variant
    INITIATOR_CODON_VARIANT = "initiator_codon_variant"
    #: Start retained
    START_RETAINED = "start_retained"
    #: Stop retained variant
    STOP_RETAINED_VARIANT = "stop_retained_variant"
    #: Synonymous variant
    SYNONYMOUS_VARIANT = "synonymous_variant"

    # modifier

    # skipped 3_prime_UTR_variant
    # skipped 5_prime_UTR_variant
    # skipped coding_sequence_variant
    # skipped conserved_intergenic_variant
    # skipped conserved_intron_variant
    #: Downstream gene variant
    DOWNSTREAM_GENE_VARIANT = "downstream_gene_variant"
    # skipped exon_variant
    # skipped feature_elongation
    # skipped feature_truncation
    # skipped gene_variant
    # skipped intergenic_variant
    #: Intron variant.
    INTRON_VARIANT = "intron_variant"
    # skipped mature_miRNA_variant
    # skipped miRNA
    # skipped NMD_transcript_variant
    #: Non-coding transcript exon variant
    NON_CODING_TRANSCRIPT_EXON_VARIANT = "non_coding_transcript_exon_variant"
    #: Non-coding transcript intron variant
    NON_CODING_TRANSCRIPT_INTRON_VARIANT = "non_coding_transcript_intron_variant"
    #: 5' UTR variant
    FIVE_PRIME_UTR_VARIANT = "5_prime_UTR_variant"
    #: Coding sequence variant
    CODING_SEQUENCE_VARIANT = "coding_sequence_variant"
    # skipped regulatory_region_amplification
    # skipped regulatory_region_variant
    # skipped TF_binding_site_variant
    # skipped TFBS_amplification
    # skipped transcript_amplification
    # skipped transcript_variant
    #: Upstream gene variant
    UPSTREAM_GENE_VARIANT = "upstream_gene_variant"

    #: EXTRA, not directly written by Mehari: 3' UTR variant + exon variant
    THREE_PRIME_UTR_VARIANT_EXON_VARIANT = "3_prime_UTR_variant-exon_variant"
    #: EXTRA, not directly written by Mehari: 5' UTR variant + exon variant
    FIVE_PRIME_UTR_VARIANT_EXON_VARIANT = "5_prime_UTR_variant-exon_variant"
    #: EXTRA, not directly written by Mehari: 3' UTR variant + intron_variant
    THREE_PRIME_UTR_VARIANT_INTRON_VARIANT = "3_prime_UTR_variant-intron_variant"
    #: EXTRA, not directly written by Mehari: 5' UTR variant + intron_variant
    FIVE_PRIME_UTR_VARIANT_INTRON_VARIANT = "5_prime_UTR_variant-intron_variant"
```

Comment (holtgrewe) [reply protos/varfish/v1/query_seqvars.proto:9]:
```
// Enumeration for `Transcript::biotype`.
enum TranscriptBiotype {
    // unknown
    TRANSCRIPT_BIOTYPE_UNKNOWN = 0;
    // Coding transcript.
    TRANSCRIPT_BIOTYPE_CODING = 1;
    // Non-coding transcript.
    TRANSCRIPT_BIOTYPE_NON_CODING = 2;
}
```

File: protos/varfish/v1/query_seqvars.proto [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    RegulatoryRegionAmplification = 42;
+    RegulatoryRegionVariant = 43;
+    TfBindingSiteVariant = 44;
+    TfbsAmplification = 45;
+    TranscriptAmplification = 46;
+    TranscriptVariant = 47;
+    UpstreamGeneVariant = 48;
+}
+
+// Enumeration for recessive mode queries.
```
Comment (holtgrewe) [inline protos/varfish/v1/query_seqvars.proto:66]: Should be as follows, same below.

```
// Unknown recessive mode.
RECESSIVE_MODE_UNKNOWN = 0;
// Generic recessive mode (homozygous or compound heterozygous recessive).
RECESSIVE_MODE_RECESSIVE = 1;
// Compound heterozygous recessive mode.
RECESSIVE_MODE_COMPOUND_RECESSIVE = 2;
```

File: protos/varfish/v1/query_seqvars.proto [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+    optional int32 heterozygous = 3;
+    // Maximal number of in-house homozygous carriers
+    optional int32 homozygous = 4;
+    // Maximal number of in-house hemizygous carriers
+    optional int32 hemizygous = 5;
+}
+
+// TODO: Might be better to structure with an optional submessage instead of all these individual optionals unless they need to be this way for some reason
+
+message GnomadOptions {
```
Comment (holtgrewe) [inline protos/varfish/v1/query_seqvars.proto:166]: Should be called and adjusted, see change below.

```
message GnomadNuclearOptions {
    // Whether to enable filtration by 1000 Genomes.
    bool enabled = 1;
    // Maximal number of carriers
    optional int32 carriers = 2;
    // Maximal number of heterozygous carriers
    optional int32 heterozygous = 3;
    // Maximal number of homozygous carriers
    optional int32 homozygous = 4;
    // Maximal number of hemizygous carriers
    optional int32 hemizygous = 5;
}
```

Also, probably add here:

```
message GnomadMitchondrialOptions {
    // Whether to enable filtration by 1000 Genomes.
    bool enabled = 1;
    // Maximal number of carriers
    optional int32 carriers = 2;
    // Maximal number of heteroplasmic carriers.
    optional int32 heteroplasmic = 3;
    // Maximal number of homoplasmic carriers.
    optional int32 homoplasmic = 4;
    // Maximal allele frequency.
    optional float allele_frequency = 5;
}
```

File: protos/varfish/v1/query_seqvars.proto [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    bool enabled = 1;
+    // Maximal frequency in HelixMtDb
+    optional float frequency = 2;
+    // Maximal number of heterozygous carriers in HelixMtDb
+    optional int32 heteroplasmic = 3;
+    // Maximal number of homozygous carriers in HelixMtDb
+    optional int32 homoplasmic = 4;
+}
+
+message PopulationFrequencyOptions {
```
Comment (holtgrewe) [inline protos/varfish/v1/query_seqvars.proto:200]: Rather use

```protobuf
    // gnomAD-exomes filter
    GnomadNuclearOptions gnomad_exomes = 1;
    // gnomAD-genomes filter
    GnomadNuclearOptions gnomad_genomes = 2;
    // gnomAD-MT filter
    GnomadMitochondrialOptions gnomad_mt = 3;
    // HelixMtDb filter
    HelixMtDbOptions helixmtdb = 4;
```

File: protos/varfish/v1/query_seqvars.proto [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+message VariantTypeOptions {
+    // Whether to include SNVs
+    bool snv = 1;
+    // Whether to include indels
+    bool indel = 2;
+    // Whether to include MNVs
+    bool mnv = 3;
+}
+
+// TODO: I've taken the liberty of turning the Option<Vec>> into just repeated since protobuf doesn't support it and there appear to be no cases where checking for 0 length wouldn't lead to the same thing
```
Comment (holtgrewe) [inline protos/varfish/v1/query_seqvars.proto:214]: agreed

File: protos/varfish/v1/query_seqvars.proto [hunk 6]
```diff
@@ -0,0 +1,247 @@
+syntax = "proto3";
+
+package varfish.v1.seqvars;
+
+// Suggestion from emily that might be worth discussing: a few members where I think the enabled bool can be implied by the presence/nonpresence of an optional submessage  making for cleaner code & less branches
```
Comment (holtgrewe) [inline protos/varfish/v1/query_seqvars.proto:5]: Please keep as is.

File: protos/varfish/v1/query_seqvars.proto [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    optional int32 carriers = 2;
+    // Maximal number of in-house heterozygous carriers
+    optional int32 heterozygous = 3;
+    // Maximal number of in-house homozygous carriers
+    optional int32 homozygous = 4;
+    // Maximal number of in-house hemizygous carriers
+    optional int32 hemizygous = 5;
+}
+
+// TODO: Might be better to structure with an optional submessage instead of all these individual optionals unless they need to be this way for some reason
```
Comment (holtgrewe) [inline protos/varfish/v1/query_seqvars.proto:164]: Please implement as specified below.

File: protos/varfish/v1/query_seqvars.proto [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+    // Index in comp. het. recessive inheritance
+    ComphetIndex = 6;
+    // Index in recessive inheritance
+    RecessiveIndex = 7;
+    // Parent in recessive inheritance
+    RecessiveParent = 8;
+}
+
+// Quality settings for one sample.
+message QualitySettings {
```
Comment (holtgrewe) [inline protos/varfish/v1/query_seqvars.proto:105]: Please use

```
// Quality settings for one sample.
message SampleQualitySettings {
    // Name of the sample filtered for
    string sample = 1;
    // Enable filter for sample.
    boolean enabled = 2;
    // .. remaining code, fields indexes shifted + 2
```

File: protos/varfish/v1/query_seqvars.proto [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+// Enumeration for recessive mode queries.
+enum RecesssiveMode {
+    Recessive = 0;
+    CompoundRecessive = 1;
+}
+
+// Choices for failing quality thresholds on genotypes.
+enum FailChoice {
```
Comment (holtgrewe) [inline protos/varfish/v1/query_seqvars.proto:73]: Can be removed

File: protos/varfish/v1/query_seqvars.proto [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+    optional int32 dp_hom = 2;
+    // Minimal genotype quality
+    optional int32 gq = 3;
+    // Minimal allele balance for het. variants
+    optional float ab = 4;
+    // Minimal number of alternate reads
+    optional int32 ad = 5;
+    // Maximal number of alternate reads
+    optional int32 ad_max = 6;
+    // Behaviour on failing quality thresholds
```
Comment (holtgrewe) [inline protos/varfish/v1/query_seqvars.proto:118]: Remove fail `FailChoice fail`

**Review by holtgrewe — CHANGES_REQUESTED** [review]
Reviewed the probotuf definition file.

File: protos/varfish/v1/query_seqvars.proto (hunk 3 above)
Comment (jakibaki) [reply protos/varfish/v1/query_seqvars.proto:166]: Is that missing gnomad_exomes.frequency? (freqency.rs for example wants to access `exomes_frequency` but can't)

Comment (holtgrewe) [reply protos/varfish/v1/query_seqvars.proto:166]: Yes, `GnomadNuclearOptions` must have

```
    // Maximal allele frequency.
    optional float allele_frequency = 6;
```

File: protos/varfish/v1/query_seqvars.proto (hunk 9 above)
Comment (jakibaki) [reply protos/varfish/v1/query_seqvars.proto:73]: There's a fair bit of code referencing this. Should I make all that code treat it as the default of ignore?

Comment (holtgrewe) [reply protos/varfish/v1/query_seqvars.proto]: Default is drop variant.

Comment (jakibaki) [reply protos/varfish/v1/query_seqvars.proto]: Oki!

Comment (jakibaki) [reply protos/varfish/v1/query_seqvars.proto]:
```
#[serde(rename_all = "kebab-case")]
pub enum FailChoice {
    /// Ignore failure.
    #[default]
    Ignore,
    /// Drop whole variant.
    #[serde(rename = "drop-variant")]
    Drop,
    /// Interpret as "no-call".
    NoCall,
}
```
(for the record in the rust code rn it's set as ignore by default which is why I made that assumption)

Comment (jakibaki) [reply protos/varfish/v1/query_seqvars.proto]: https://github.com/varfish-org/varfish-server-worker/blob/402d10d98aa9106734db16fcbecd3a193ded24fd/src/seqvars/query/interpreter/quality.rs#L161 how should I handle the (now) failing test cases here now that there's no way to specify the FailChoice?  Should I just remove the ones without drop?

Comment (jakibaki) [reply protos/varfish/v1/query_seqvars.proto]: (as well as all code referencing other failchoices?)

**Review by holtgrewe — APPROVED** [review]

**Comment by jakibaki** [comment]
I think this is ready for a final round of review then I can properly split this up into the two PRs (if that's still wanted).

**Pull request closed without merging**